src/app_engine/message_page.py

```diff
@@ -43,10 +44,19 @@ def send_message_to_collider(self, room_id, client_id, message):
       return
     self.write_response(constants.RESPONSE_SUCCESS)
 
-  def post(self, room_id, client_id):
-    message_json = self.request.body
+  def post(self, room_id, client_id = None):
+    message = self.request.body
+    if client_id is None:
+      # Check for client_id in body.
+      msg = util.get_message_from_json(self.request.body)
+      if not util.has_msg_fields(msg, ((constants.PARAM_USER_GCM_ID, basestring), (constants.PARAM_MESSAGE, basestring))):
+        self.write_response(constants.RESPONSE_INVALID_ARGUMENT)
+        return
+      client_id = msg[constants.PARAM_USER_GCM_ID]
+      message = msg[constants.PARAM_MESSAGE]
+
     result = room_module.save_message_from_client(
```

Comment:
Not introduced in this CL, but we have UNKNOWN_ROOM/UNKNOWN_CLIENT, duplicated with INVALID_ROOM/INVALID_USER. Maybe this is a good chance to unify them?